Userland/DevTools/UserspaceEmulator/SoftFPU.cpp

```diff
@@ -0,0 +1,1744 @@
+        // FIXME: Respect shadow values
+        auto f64 = bit_cast<double>(new_f64.value());
+        // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+        fpu_set(0, f64 / fpu_get(0));
+    }
+}
+void SoftFPU::FDIVRP(const X86::Instruction& insn)
+{
+    VERIFY(insn.modrm().is_register());
+    // FIXME: Raise IA on + infinity / +-infinity, +-0 / +-0, raise Z on finite / +-0
+    fpu_set(insn.modrm().register_index(), fpu_get(0) / fpu_get(insn.modrm().register_index()));
+    fpu_pop();
+}
+
+void SoftFPU::FIDIV_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m16int = (i16)insn.modrm().read16(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    // FIXME: Raise IA on 0 / _=0, raise Z on finite / +-0
+    fpu_set(0, fpu_get(0) / (long double)m16int);
+}
+void SoftFPU::FIDIV_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = (i32)insn.modrm().read32(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    // FIXME: Raise IA on 0 / _=0, raise Z on finite / +-0
+    fpu_set(0, fpu_get(0) / (long double)m32int);
+}
+
+void SoftFPU::FIDIVR_RM16(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m16int = (i16)insn.modrm().read16(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    // FIXME: Raise IA on 0 / _=0, raise Z on finite / +-0
+    fpu_set(0, (long double)m16int / fpu_get(0));
+}
+void SoftFPU::FIDIVR_RM32(const X86::Instruction& insn)
+{
+    VERIFY(!insn.modrm().is_register());
+    auto m32int = (i32)insn.modrm().read32(m_cpu, insn).value();
+    // FIXME: Respect shadow values
+    // FIXME: Raise IA on 0 / _=0, raise Z on finite / +-0
+    fpu_set(0, (long double)m32int / fpu_get(0));
+}
+
+void SoftFPU::FPREM(const X86::Instruction&)
+{
+    // FIXME: There are some exponent shenanigans supposed to be here
+    long double top = fpu_get(0);
+    long double one = fpu_get(1);
+
+    int Q = static_cast<int>(truncl(top / one));
+    top = top - (one * Q);
+    set_c2(0);
+    set_c1(Q & 1);
+    set_c3((Q >> 1) & 1);
+    set_c0((Q >> 2) & 1);
+
+    fpu_set(0, top);
+}
+void SoftFPU::FPREM1(const X86::Instruction&)
+{
+    // FIXME: There are some exponent shenanigans supposed to be here
+    long double top = fpu_get(0);
+    long double one = fpu_get(1);
+
+    int Q = static_cast<int>(roundl(top / one));
+    top = top - (one * Q);
+    set_c2(0);
+    set_c1(Q & 1);
+    set_c3((Q >> 1) & 1);
+    set_c0((Q >> 2) & 1);
+
+    fpu_set(0, top);
+}
+void SoftFPU::FABS(const X86::Instruction&)
+{
+    set_c1(0);
+    fpu_set(0, __builtin_fabsl(fpu_get(0)));
+}
+void SoftFPU::FCHS(const X86::Instruction&)
+{
+    set_c1(0);
+    fpu_set(0, -fpu_get(0));
+}
+
+void SoftFPU::FRNDINT(const X86::Instruction&)
+{
+    auto res = fpu_round<long double>(fpu_get(0));
+    if (auto rnd = (res - fpu_get(0))) {
```

Comment:
Bit c1 is only set if it's not an exact conversion? Spec says

> Set if result was rounded up; cleared otherwise.

Should remove the top if and convert the lower to something like "fpu_get(0) > res".

Reply:
I realize we have round_checked for that and replaced the block
and changed the behavior of that helper to be actually correct as you described